models/spring-ai-openai/src/main/java/org/springframework/ai/openai/OpenAiChatModel.java

```diff
@@ -143,135 +142,134 @@ public ChatResponse call(Prompt prompt) {
-							}
-							String finish = (choice.finishReason() != null ? choice.finishReason().name() : "");
-							var generation = new Generation(choice.message().content(),
-									Map.of("id", id, "role", roleMap.getOrDefault(id, ""), "finishReason", finish));
-							if (choice.finishReason() != null) {
-								generation = generation.withGenerationMetadata(
-										ChatGenerationMetadata.from(choice.finishReason().name(), null));
-							}
-							return generation;
-						}).toList();
-
-						if (chatCompletion2.usage() != null) {
-							return new ChatResponse(generations, OpenAiChatResponseMetadata.from(chatCompletion2));
-						}
-						else {
-							return new ChatResponse(generations);
+					List<Generation> generations = chatCompletion2.choices().stream().map(choice -> {
+						if (choice.message().role() != null) {
+							roleMap.putIfAbsent(id, choice.message().role().name());
 						}
+			// @formatter:off
+						Map<String, Object> metadata = Map.of(
+								"id", chatCompletion2.id(),
+								"role", roleMap.getOrDefault(id, ""),
+								"finishReason", choice.finishReason() != null ? choice.finishReason().name() : "");
+						// @formatter:on
+						return buildGeneration(choice, metadata);
+					}).toList();
+
+					if (chatCompletion2.usage() != null) {
+						return new ChatResponse(generations, OpenAiChatResponseMetadata.from(chatCompletion2));
 					}
-					catch (Exception e) {
-						logger.error("Error processing chat completion", e);
-						return new ChatResponse(List.of());
+					else {
+						return new ChatResponse(generations);
 					}
+				}
+				catch (Exception e) {
+					logger.error("Error processing chat completion", e);
+					return new ChatResponse(List.of());
+				}
+
+			}));
 
-				});
-			});
+		return chatResponse.flatMap(response -> {
+			if (isToolCall(response, OpenAiApi.ChatCompletionFinishReason.TOOL_CALLS.name())) {
+				var toolCallConversation = handleToolCalls(prompt, response);
+				// Recursively call the stream method with the tool call message
+				// conversation that contains the call responses.
+				return this.stream(new Prompt(toolCallConversation, prompt.getOptions()));
+			}
+			else {
+				return Flux.just(response);
+			}
 		});
 	}
 
-	private List<Message> handleToolCallRequests(List<Message> previousMessages, ChatCompletion chatCompletion) {
-
-		ChatCompletionMessage nativeAssistantMessage = this.extractAssistantMessage(chatCompletion);
-
-		List<AssistantMessage.ToolCall> assistantToolCalls = nativeAssistantMessage.toolCalls()
-			.stream()
-			.map(toolCall -> new AssistantMessage.ToolCall(toolCall.id(), "function", toolCall.function().name(),
-					toolCall.function().arguments()))
-			.toList();
-
-		AssistantMessage assistantMessage = new AssistantMessage(nativeAssistantMessage.content(), Map.of(),
-				assistantToolCalls);
+	private static Generation buildGeneration(Choice choice, Map<String, Object> metadata) {
```

Comment:
This part was duplicated in both "call()" and "stream()". I moved it out to avoid adding more duplications with the tool call handling.